stock_picking_dispatch/models/stock_move.py

```diff
@@ -0,0 +1,56 @@
+# -*- coding: utf-8 -*-
+# © 2012-2014 Alexandre Fayolle, Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import logging
+from openerp import fields, models
+
+_logger = logging.getLogger(__name__)
+
+
+class StockMove(models.Model):
+    _inherit = 'stock.move'
+    dispatch_id = fields.Many2one(
+        'picking.dispatch', 'Dispatch', index=True, copy=False,
+        help='who this move is dispatched to'
+    )
+
+    def action_cancel(self):
+        """
+        in addition to what the method in the parent class does,
+        cancel the dispatches for which all moves where cancelled
+        """
+        _logger.debug('cancel stock.moves %s', self.ids)
+        status = super(StockMove, self).action_cancel()
+
+        if self.ids:
+            dispatches = self.env['picking.dispatch']
+            for move in self:
+                if move.dispatch_id:
+                    dispatches |= move.dispatch_id
```

Comment:
Here you can probably remove the `for` loop by using `mapped` with

    dispatches = self.mapped('dispatch_id')